lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -2342,6 +2483,52 @@ instance DecodeAddress t => FromJSON (PostTransactionData t) where
 instance EncodeAddress t => ToJSON (PostTransactionData t) where
     toJSON = genericToJSON defaultRecordTypeOptions
 
+instance DecodeAddress t => FromJSON (ApiPaymentDestination t) where
+    parseJSON obj = parseAll <|> parseAddrs
+      where
+        parseAll = ApiPaymentAll <$> parseJSON obj
+        parseAddrs = ApiPaymentAddresses <$> parseJSON obj
+
+instance EncodeAddress t => ToJSON (ApiPaymentDestination t) where
+    toJSON (ApiPaymentAddresses addrs) = toJSON addrs
+    toJSON (ApiPaymentAll addrs) = toJSON addrs
+
+instance DecodeAddress t => FromJSON (ApiConstructTransactionData t) where
+    parseJSON = genericParseJSON defaultRecordTypeOptions
+instance EncodeAddress t => ToJSON (ApiConstructTransactionData t) where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+instance ToJSON ApiValidityBound where
+    toJSON ApiValidityBoundUnspecified = String "unspecified"
+    toJSON (ApiValidityBoundAsTimeFromNow from) = toJSON from
+    toJSON (ApiValidityBoundAsSlot sl) = toJSON sl
+instance FromJSON ApiValidityBound where
+    parseJSON obj = processString obj <|> processObject obj
+      where
+        processString = withText "ApiValidityBound string" $ \str ->
+            if str == "unspecified" then
+                pure ApiValidityBoundUnspecified
+            else
+                fail "invalid string of ApiValidityBound"
+        processObject = withObject "ApiValidityBound object" $ \o -> do
```

Comment:
It should be possible try parsing the Quantities directly without testing the "unit" field.